docs/manpages/vfs_fileid.8
'\" t
.\"     Title: vfs_fileid
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 12/15/2022
.\"    Manual: System Administration tools
.\"    Source: Samba 4.15.13
.\"  Language: English
.\"
.TH "VFS_FILEID" "8" "12/15/2022" "Samba 4\&.15\&.13" "System Administration tools"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
vfs_fileid \- Generates file_id structs with unique device id values for cluster setups
.SH "SYNOPSIS"
.HP \w'\ 'u
vfs objects = fileid
.SH "DESCRIPTION"
.PP
This VFS module is part of the
\fBsamba\fR(7)
suite\&.
.PP
Samba uses file_id structs to uniquely identify files for locking purpose\&. By default the file_id contains the device and inode number returned by the
stat()
system call\&. As the file_id is a unique identifier of a file, it must be the same on all nodes in a cluster setup\&. This module overloads the
SMB_VFS_FILE_ID_CREATE()
operation and generates the device number based on the configured algorithm (see the "fileid:algorithm" option)\&.
.PP
When using the fsname or fsid algorithm a
stat()
and
statfs()
call is required for all mounted file systems to generate the file_id\&. If e\&.g\&. an NFS file system is unresponsive such a call might block and the smbd process will become unresponsive\&. Use the "fileid:fstype deny", "fileid:fstype allow", "fileid:mntdir deny", or "fileid:mntdir allow" options to ignore potentially unresponsive file systems\&.
.SH "OPTIONS"
.PP
fileid:algorithm = ALGORITHM
.RS 4
Available algorithms are
fsname,
fsname_nodirs,
fsid
and
hostname\&. The default value is
fsname\&.
.sp
The
fsname
algorithm generates device id by hashing the kernel device name\&.
.sp
The
fsname_nodirs
algorithm generates device id by hashing the kernel device name for files and by hashing the hostname for directories\&. This can be used to deliberately break lock coherency for directories in a cluster\&.
.sp
The
fsid
algorithm generates the device id from the
f_fsid
returned from the
statfs()
syscall\&.
.sp
The
hostname
algorithm generates device id by hashing the hostname\&. This can be used to deliberately break lock coherency in a cluster\&.
.sp
The
fsname_norootdir
algorithm generates device ids by hashing the kernel device name, except for the root directory of shares where it will use the hostname algorithm\&. This can be used to deliberately break lock coherency in a cluster for the root directory of a share\&.
.sp
The
fsname_norootdir_ext
algorithm generates device ids by hashing the kernel device name, except for the root directory of shares where it will use the hostname algorithm\&. Additionally it generates an extid based on the process pid\&. This can be used to deliberately break lock coherency between all smbd processes in the whole cluster for the root directory of a share\&.
.RE
.PP
fileid:mapping = ALGORITHM
.RS 4
This option is the legacy version of the
fileid:algorithm
option, which was used in earlier versions of fileid mapping feature in custom Samba 3\&.0 versions\&.
.RE
.PP
fileid:fstype deny = LIST
.RS 4
List of file system types to be ignored for file_id generation\&.
.RE
.PP
fileid:fstype allow = LIST
.RS 4
List of file system types to be allowed for file_id generation\&. If this option is set, file system types not listed here are ignored\&.
.RE
.PP
fileid:mntdir deny = LIST
.RS 4
List of file system mount points to be ignored for file_id generation\&.
.RE
.PP
fileid:mntdir allow = LIST
.RS 4
List of file system mount points to be allowed for file_id generation\&. If this option is set, file system mount points not listed here are ignored\&.
.RE
.PP
fileid:nolockinode
.RS 4
This option triggers use of the fileid hostname algorithm for the configured inode which can be used to deliberately break lock coherency for the corresponding file or directory in a cluster\&.
.RE
.SH "EXAMPLES"
.PP
Usage of the
fileid
module with the
fsid
algorithm:
.sp
.if n \{\
.RS 4
.\}
.nf
        \fI[global]\fR
	\m[blue]\fBvfs objects = fileid\fR\m[]
	\m[blue]\fBfileid:algorithm = fsid\fR\m[]
.fi
.if n \{\
.RE
.\}
.SH "VERSION"
.PP
This man page is part of version 4\&.15\&.13 of the Samba suite\&.
.SH "AUTHOR"
.PP
The original Samba software and related utilities were created by Andrew Tridgell\&. Samba is now developed by the Samba Team as an Open Source project similar to the way the Linux kernel is developed\&.


debian/NEWS
samba (2:4.4.1+dfsg-1) experimental; urgency=medium

    This Samba security addresses both Denial of Service and Man in
    the Middle vulnerabilities.

    Both of these changes implement new smb.conf options and a number
    of stricter behaviours to prevent Man in the Middle attacks on our
    network services, as a client and as a server.

    Between these changes, compatibility with a large number of older
    software versions has been lost in the default configuration.

    See the release notes in WHATNEW.txt for more information.


    Here are some additional hints how to work around the new stricter default behaviors:

    * As an AD DC server, only Windows 2000 and Samba 3.6 and above as
      a domain member are supported out of the box. Other smb file
      servers as domain members are also fine out of the box.

    * As an AD DC server, with default setting of "ldap server require
      strong auth", LDAP clients connecting over ldaps:// or START_TLS
      will be allowed to perform simple LDAP bind only.

      The preferred configuration for LDAP clients is to use SASL
      GSSAPI directly over ldap:// without using ldaps:// or
      START_TLS.

      To use LDAP with START_TLS and SASL GSSAPI (either Kerberos or
      NTLMSSP) sign/seal protection must be used by the client and
      server should be configured with "ldap server require strong
      auth = allow_sasl_over_tls".

      Consult OpenLDAP documentation how to set sign/seal protection
      in ldap.conf.

      For SSSD client configured with "id_provider = ad" or
      "id_provider = ldap" with "auth_provider = krb5", see
      sssd-ldap(5) manual for details on TLS session handling.

    * As a File Server, compatibility with the Linux Kernel cifs
      client depends on which configuration options are selected, please
      use "sec=krb5(i)" or "sec=ntlmssp(i)", not "sec=ntlmv2".

    * As a file or printer client and as a domain member, out of the
      box compatibility with Samba less than 4.0 and other SMB/CIFS
      servers, depends on support for SMB signing or SMB2 on the
      server, which is often disabled or absent. You may need to
      adjust the "client ipc signing" to "no" in these cases.

    * In case of an upgrade from versions before 4.2.0, you might run
      into problems as a domain member. The out of the box compatibility
      with Samba 3.x domain controllers requires NETLOGON features only
      available in Samba 3.2 and above.

    However, all of these can be worked around by setting smb.conf
    options in Samba, see WHATSNEW.txt the 4.2.0 release notes at
    https://www.samba.org/samba/history/samba-4.2.0.html and the Samba
    wiki for details, workarounds and suggested security-improving
    changes to these and other software packages.


    Suggested further improvements after patching:

    It is recommended that administrators set these additional options,
    if compatible with their network environment:

        server signing = mandatory
        ntlm auth = no

    Without "server signing = mandatory", Man in the Middle attacks
    are still possible against our file server and
    classic/NT4-like/Samba3 Domain controller. (It is now enforced on
    Samba's AD DC.) Note that this has heavy impact on the file server
    performance, so you need to decide between performance and
    security. These Man in the Middle attacks for smb file servers are
    well known for decades.

    Without "ntlm auth = no", there may still be clients not using
    NTLMv2, and these observed passwords may be brute-forced easily using
    cloud-computing resources or rainbow tables.

 -- Andrew Bartlett <abartlet+debian@catalyst.net.nz>  Tue, 12 Apr 2016 16:18:57 +1200

samba (2:4.0.10+dfsg-3) unstable; urgency=low

    The SWAT package is no longer available.

    Upstream support for SWAT (Samba Web Administration Tool) was removed in
    samba 4.1.0. As a result, swat is no longer shipped in the Debian Samba
    packages. Unfortunately, there is currently no replacement.

    Details why SWAT has been removed upstream can be found on the
    samba-technical mailing list:

    https://lists.samba.org/archive/samba-technical/2013-February/090572.html

 -- Ivo De Decker <ivo.dedecker@ugent.be>  Tue, 22 Oct 2013 07:52:54 +0200

samba (2:3.6.5-2) unstable; urgency=low

    NSS modules have been split out from libpam-winbind to
    libnss-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libnss-winbind package after upgrading
    from former versions of winbind (for instance from squeeze) or
    from former versions of libpam-winbind.

 -- Christian Perrier <bubulle@debian.org>  Mon, 07 May 2012 22:16:32 +0200

samba (2:3.5.11~dfsg-3) unstable; urgency=low

    PAM modules and NSS modules have been split out from the winbind
    package into libpam-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libpam-winbind package after upgrading
    from former versions of winbind (for instance from squeeze)

 -- Steve Langasek <vorlon@debian.org>  Fri, 21 Oct 2011 20:00:13 +0000

samba (2:3.4.0-1) unstable; urgency=low

    Default passdb backend changed in samba 3.4.0 and above
  
    Beginning with samba 3.4.0, the default setting for "passdb
    backend" changed from "smbpasswd" to "tdbsam".
    
    If your smb.conf file does not have an explicit mention of
    "passdb backend" when upgrading from pre-3.4.0 versions of
    samba, it is likely that users will no longer be able to
    authenticate.
    
    As a consequence of all this, if you're upgrading from lenny
    and have no setting of "passdb backend" in smb.conf, you MUST
    add "passdb backend = smbpasswd" in order to keep your samba
    server's behaviour.
    
    As Debian packages of samba explicitly set "passdb backend = tdbsam"
    by default since etch, very few users should need to modify their

source3/modules/vfs_fileid.c
/*
 * VFS module to alter the algorithm to calculate
 * the struct file_id used as key for the share mode
 * and byte range locking db's.
 *
 * Copyright (C) 2007, Stefan Metzmacher
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

#include "includes.h"
#include "smbd/smbd.h"
#include "system/filesys.h"

static int vfs_fileid_debug_level = DBGC_VFS;

#undef DBGC_CLASS
#define DBGC_CLASS vfs_fileid_debug_level

struct fileid_mount_entry {
	SMB_DEV_T device;
	const char *mnt_fsname;
	fsid_t fsid;
	uint64_t devid;
};

struct fileid_nolock_inode {
	dev_t dev;
	ino_t ino;
};

struct fileid_handle_data {
	struct vfs_handle_struct *handle;
	struct file_id (*mapping_fn)(struct fileid_handle_data *data,
				     const SMB_STRUCT_STAT *sbuf);
	char **fstype_deny_list;
	char **fstype_allow_list;
	char **mntdir_deny_list;
	char **mntdir_allow_list;
	unsigned num_mount_entries;
	struct fileid_mount_entry *mount_entries;
	struct {
		bool force_all_inodes;
		bool force_all_dirs;
		uint64_t extid;
		size_t num_inodes;
		struct fileid_nolock_inode *inodes;
	} nolock;
};

/* check if a mount entry is allowed based on fstype and mount directory */
static bool fileid_mount_entry_allowed(struct fileid_handle_data *data,
				       struct mntent *m)
{
	int i;
	char **fstype_deny = data->fstype_deny_list;
	char **fstype_allow = data->fstype_allow_list;
	char **mntdir_deny = data->mntdir_deny_list;
	char **mntdir_allow = data->mntdir_allow_list;

	if (fstype_deny != NULL) {
		for (i = 0; fstype_deny[i] != NULL; i++) {
			if (strcmp(m->mnt_type, fstype_deny[i]) == 0) {
				return false;
			}
		}
	}
	if (fstype_allow != NULL) {
		for (i = 0; fstype_allow[i] != NULL; i++) {
			if (strcmp(m->mnt_type, fstype_allow[i]) == 0) {
				break;
			}
		}
		if (fstype_allow[i] == NULL) {
			return false;
		}
	}
	if (mntdir_deny != NULL) {
		for (i=0; mntdir_deny[i] != NULL; i++) {
			if (strcmp(m->mnt_dir, mntdir_deny[i]) == 0) {
				return false;
			}
		}
	}
	if (mntdir_allow != NULL) {
		for (i=0; mntdir_allow[i] != NULL; i++) {
			if (strcmp(m->mnt_dir, mntdir_allow[i]) == 0) {
				break;
			}
		}
		if (mntdir_allow[i] == NULL) {
			return false;
		}
	}
	return true;
}


/* load all the mount entries from the mtab */
static void fileid_load_mount_entries(struct fileid_handle_data *data)
{
	FILE *f;
	struct mntent *m;

	data->num_mount_entries = 0;
	TALLOC_FREE(data->mount_entries);

#ifndef __OS2__
	f = setmntent("/etc/mtab", "r");
#endif
	if (!f) return;

#ifndef __OS2__
	while ((m = getmntent(f))) {
		struct stat st;
		struct statfs sfs;
		struct fileid_mount_entry *cur;
		bool allowed;

		allowed = fileid_mount_entry_allowed(data, m);
		if (!allowed) {
			DBG_DEBUG("skipping mount entry %s\n", m->mnt_dir);
			continue;
		}
		if (stat(m->mnt_dir, &st) != 0) continue;
		if (statfs(m->mnt_dir, &sfs) != 0) continue;

		if (strncmp(m->mnt_fsname, "/dev/", 5) == 0) {
			m->mnt_fsname += 5;
		}

		data->mount_entries = talloc_realloc(data,
							   data->mount_entries,
							   struct fileid_mount_entry,
							   data->num_mount_entries+1);
		if (data->mount_entries == NULL) {
			goto nomem;
		}

		cur = &data->mount_entries[data->num_mount_entries];
		cur->device	= st.st_dev;
		cur->mnt_fsname = talloc_strdup(data->mount_entries,
						m->mnt_fsname);
		if (!cur->mnt_fsname) goto nomem;
		cur->fsid	= sfs.f_fsid;
		cur->devid	= (uint64_t)-1;

		data->num_mount_entries++;
	}
	endmntent(f);
#endif
	return;
	
nomem:
	if (f) endmntent(f);

	data->num_mount_entries = 0;
	TALLOC_FREE(data->mount_entries);

	return;
}

/* find a mount entry given a dev_t */
static struct fileid_mount_entry *fileid_find_mount_entry(struct fileid_handle_data *data,
							  SMB_DEV_T dev)
{
	unsigned i;

	if (data->num_mount_entries == 0) {
		fileid_load_mount_entries(data);
	}
	for (i=0;i<data->num_mount_entries;i++) {
		if (data->mount_entries[i].device == dev) {
			return &data->mount_entries[i];
		}
	}
	/* 2nd pass after reloading */
	fileid_load_mount_entries(data);
	for (i=0;i<data->num_mount_entries;i++) {
		if (data->mount_entries[i].device == dev) {
			return &data->mount_entries[i];
		}
	}	
	return NULL;
}


/* a 64 bit hash, based on the one in tdb */
static uint64_t fileid_uint64_hash(const uint8_t *s, size_t len)
{
	uint64_t value;	/* Used to compute the hash value.  */
	uint32_t i;	/* Used to cycle through random values. */

	/* Set the initial value from the key size. */
	for (value = 0x238F13AFLL * len, i=0; i < len; i++)
		value = (value + (((uint64_t)s[i]) << (i*5 % 24)));

	return (1103515243LL * value + 12345LL);
}

/* a device mapping using a fsname */
static uint64_t fileid_device_mapping_fsname(struct fileid_handle_data *data,
					     const SMB_STRUCT_STAT *sbuf)
{
	struct fileid_mount_entry *m;

	m = fileid_find_mount_entry(data, sbuf->st_ex_dev);
	if (!m) return sbuf->st_ex_dev;

	if (m->devid == (uint64_t)-1) {
		m->devid = fileid_uint64_hash((const uint8_t *)m->mnt_fsname,
					      strlen(m->mnt_fsname));
	}

	return m->devid;
}

static struct file_id fileid_mapping_fsname(struct fileid_handle_data *data,
					    const SMB_STRUCT_STAT *sbuf)
{
	struct file_id id = { .inode = sbuf->st_ex_ino, };

	id.devid = fileid_device_mapping_fsname(data, sbuf);

	return id;
}

/* a device mapping using a hostname */
static uint64_t fileid_device_mapping_hostname(struct fileid_handle_data *data,
					       const SMB_STRUCT_STAT *sbuf)
{
	char hostname[HOST_NAME_MAX+1];
	char *devname = NULL;
	uint64_t id;
	size_t devname_len;
	int rc;

	rc = gethostname(hostname, HOST_NAME_MAX+1);
	if (rc != 0) {
		DBG_ERR("gethostname failed\n");
		return UINT64_MAX;
	}

	devname = talloc_asprintf(talloc_tos(), "%s%ju",
				  hostname, (uintmax_t)sbuf->st_ex_dev);
	if (devname == NULL) {
		DBG_ERR("talloc_asprintf failed\n");
		return UINT64_MAX;
	}
	devname_len = talloc_array_length(devname) - 1;

	id = fileid_uint64_hash((uint8_t *)devname, devname_len);

	TALLOC_FREE(devname);

	return id;
}

static struct file_id fileid_mapping_hostname(struct fileid_handle_data *data,
					      const SMB_STRUCT_STAT *sbuf)
{
	struct file_id id = { .inode = sbuf->st_ex_ino, };

	id.devid = fileid_device_mapping_hostname(data, sbuf);

	return id;
}

static bool fileid_is_nolock_inode(struct fileid_handle_data *data,
				   const SMB_STRUCT_STAT *sbuf)
{
	size_t i;

	if (data->nolock.force_all_inodes) {
		return true;
	}

	if (S_ISDIR(sbuf->st_ex_mode) && data->nolock.force_all_dirs) {
		return true;
	}

	/*
	 * We could make this a binary search over an sorted array,
	 * but for now we keep things simple.
	 */

	for (i=0; i < data->nolock.num_inodes; i++) {
		if (data->nolock.inodes[i].ino != sbuf->st_ex_ino) {
			continue;
		}

		if (data->nolock.inodes[i].dev == 0) {
			/*
			 * legacy "fileid:nolockinode"
			 * handling ignoring dev
			 */
			return true;
		}

		if (data->nolock.inodes[i].dev != sbuf->st_ex_dev) {
			continue;
		}

		return true;
	}

	return false;
}

static int fileid_add_nolock_inode(struct fileid_handle_data *data,
				   const SMB_STRUCT_STAT *sbuf)
{
	bool exists = fileid_is_nolock_inode(data, sbuf);
	struct fileid_nolock_inode *inodes = NULL;

	if (exists) {
		return 0;
	}

	inodes = talloc_realloc(data, data->nolock.inodes,
				struct fileid_nolock_inode,
				data->nolock.num_inodes + 1);
	if (inodes == NULL) {
		return -1;
	}

	inodes[data->nolock.num_inodes] = (struct fileid_nolock_inode) {
		.dev = sbuf->st_ex_dev,
		.ino = sbuf->st_ex_ino,
	};
	data->nolock.inodes = inodes;
	data->nolock.num_inodes += 1;

	return 0;
}

static uint64_t fileid_mapping_nolock_extid(uint64_t max_slots)
{
	char buf[8+4+HOST_NAME_MAX+1] = { 0, };
	uint64_t slot = 0;
	uint64_t id;
	int rc;

	if (max_slots > 1) {
		slot = getpid() % max_slots;
	}

	PUSH_LE_U64(buf, 0, slot);
	PUSH_LE_U32(buf, 8, get_my_vnn());

	rc = gethostname(&buf[12], HOST_NAME_MAX+1);
	if (rc != 0) {
		DBG_ERR("gethostname failed\n");
		return UINT64_MAX;
	}

	id = fileid_uint64_hash((uint8_t *)buf, ARRAY_SIZE(buf));

	return id;
}

/* device mapping functions using a fsid */
static uint64_t fileid_device_mapping_fsid(struct fileid_handle_data *data,
					   const SMB_STRUCT_STAT *sbuf)
{
	struct fileid_mount_entry *m;

	m = fileid_find_mount_entry(data, sbuf->st_ex_dev);
	if (!m) return sbuf->st_ex_dev;

	if (m->devid == (uint64_t)-1) {
		if (sizeof(fsid_t) > sizeof(uint64_t)) {
			m->devid = fileid_uint64_hash((uint8_t *)&m->fsid,
						      sizeof(m->fsid));
		} else {
			union {
				uint64_t ret;
				fsid_t fsid;
			} u;
			ZERO_STRUCT(u);
			u.fsid = m->fsid;
			m->devid = u.ret;
		}
	}

	return m->devid;
}

static struct file_id fileid_mapping_fsid(struct fileid_handle_data *data,
					  const SMB_STRUCT_STAT *sbuf)
{
	struct file_id id = { .inode = sbuf->st_ex_ino, };

	id.devid = fileid_device_mapping_fsid(data, sbuf);

	return id;
}

static struct file_id fileid_mapping_next_module(struct fileid_handle_data *data,
						 const SMB_STRUCT_STAT *sbuf)
{
	return SMB_VFS_NEXT_FILE_ID_CREATE(data->handle, sbuf);
}

static int get_connectpath_ino(struct vfs_handle_struct *handle,
			       const char *path,
			       SMB_STRUCT_STAT *psbuf)
{
	TALLOC_CTX *frame = talloc_stackframe();
	struct smb_filename *fname = NULL;
	const char *fullpath = NULL;
	int ret;

	if (path[0] == '/') {
		fullpath = path;
	} else {
		fullpath = talloc_asprintf(frame,
					   "%s/%s",
					   handle->conn->connectpath,
					   path);
		if (fullpath == NULL) {
			DBG_ERR("talloc_asprintf() failed\n");
			TALLOC_FREE(frame);
			return -1;
		}
	}

	fname = synthetic_smb_fname(frame,
				    fullpath,
				    NULL,
				    NULL,
				    0,
				    0);
	if (fname == NULL) {
		DBG_ERR("synthetic_smb_fname(%s) failed - %s\n",
			fullpath, strerror(errno));
		TALLOC_FREE(frame);
		return -1;
	}

	ret = SMB_VFS_NEXT_STAT(handle, fname);
	if (ret != 0) {
		DBG_ERR("stat failed for %s with %s\n",
			fullpath, strerror(errno));
		TALLOC_FREE(frame);
		return -1;
	}
	*psbuf = fname->st;

	TALLOC_FREE(frame);

	return 0;
}

static int fileid_connect(struct vfs_handle_struct *handle,
			  const char *service, const char *user)
{
	struct fileid_handle_data *data;
	const char *algorithm;
	const char **fstype_deny_list = NULL;
	const char **fstype_allow_list = NULL;
	const char **mntdir_deny_list = NULL;
	const char **mntdir_allow_list = NULL;
	ino_t nolockinode;
	uint64_t max_slots = 0;
	bool rootdir_nolock = false;
	const char **nolock_paths = NULL;
	size_t i;
	int saved_errno;
	int ret = SMB_VFS_NEXT_CONNECT(handle, service, user);

	if (ret < 0) {
		return ret;
	}

	data = talloc_zero(handle, struct fileid_handle_data);
	if (!data) {
		saved_errno = errno;
		SMB_VFS_NEXT_DISCONNECT(handle);
		DEBUG(0, ("talloc_zero() failed\n"));
		errno = saved_errno;
		return -1;
	}
	data->handle = handle;

	/*
	 * "fileid:mapping" is only here as fallback for old setups
	 * "fileid:algorithm" is the option new setups should use
	 */
	algorithm = lp_parm_const_string(SNUM(handle->conn),
					 "fileid", "mapping",
					 "fsname");
	algorithm = lp_parm_const_string(SNUM(handle->conn),
					 "fileid", "algorithm",
					 algorithm);
	if (strcmp("fsname", algorithm) == 0) {
		data->mapping_fn = fileid_mapping_fsname;
	} else if (strcmp("fsname_nodirs", algorithm) == 0) {
		data->mapping_fn = fileid_mapping_fsname;
		data->nolock.force_all_dirs = true;
	} else if (strcmp("fsid", algorithm) == 0) {
		data->mapping_fn = fileid_mapping_fsid;
	} else if (strcmp("hostname", algorithm) == 0) {
		data->mapping_fn = fileid_mapping_hostname;
		data->nolock.force_all_inodes = true;
	} else if (strcmp("fsname_norootdir", algorithm) == 0) {
		data->mapping_fn = fileid_mapping_fsname;
		rootdir_nolock = true;
	} else if (strcmp("fsname_norootdir_ext", algorithm) == 0) {
		data->mapping_fn = fileid_mapping_fsname;
		rootdir_nolock = true;
		max_slots = UINT64_MAX;
	} else if (strcmp("next_module", algorithm) == 0) {
		data->mapping_fn	= fileid_mapping_next_module;
	} else {
		SMB_VFS_NEXT_DISCONNECT(handle);
		DEBUG(0,("fileid_connect(): unknown algorithm[%s]\n", algorithm));
		return -1;
	}

	fstype_deny_list = lp_parm_string_list(SNUM(handle->conn), "fileid",
					       "fstype deny", NULL);
	if (fstype_deny_list != NULL) {
		data->fstype_deny_list = str_list_copy(data, fstype_deny_list);
		if (data->fstype_deny_list == NULL) {
			saved_errno = errno;
			DBG_ERR("str_list_copy failed\n");
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}
	}

	fstype_allow_list = lp_parm_string_list(SNUM(handle->conn), "fileid",
						"fstype allow", NULL);
	if (fstype_allow_list != NULL) {
		data->fstype_allow_list = str_list_copy(data, fstype_allow_list);
		if (data->fstype_allow_list == NULL) {
			saved_errno = errno;
			DBG_ERR("str_list_copy failed\n");
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}
	}

	mntdir_deny_list = lp_parm_string_list(SNUM(handle->conn), "fileid",
					       "mntdir deny", NULL);
	if (mntdir_deny_list != NULL) {
		data->mntdir_deny_list = str_list_copy(data, mntdir_deny_list);
		if (data->mntdir_deny_list == NULL) {
			saved_errno = errno;
			DBG_ERR("str_list_copy failed\n");
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}
	}

	mntdir_allow_list = lp_parm_string_list(SNUM(handle->conn), "fileid",
						"mntdir allow", NULL);
	if (mntdir_allow_list != NULL) {
		data->mntdir_allow_list = str_list_copy(data, mntdir_allow_list);
		if (data->mntdir_allow_list == NULL) {
			saved_errno = errno;
			DBG_ERR("str_list_copy failed\n");
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}
	}

	data->nolock.force_all_inodes = lp_parm_bool(SNUM(handle->conn),
						     "fileid", "nolock_all_inodes",
						     data->nolock.force_all_inodes);
	data->nolock.force_all_dirs = lp_parm_bool(SNUM(handle->conn),
						   "fileid", "nolock_all_dirs",
						   data->nolock.force_all_dirs);

	max_slots = lp_parm_ulonglong(SNUM(handle->conn),
				      "fileid", "nolock_max_slots",
				      max_slots);
	max_slots = MAX(max_slots, 1);

	data->nolock.extid = fileid_mapping_nolock_extid(max_slots);

	nolockinode = lp_parm_ulong(SNUM(handle->conn), "fileid", "nolockinode", 0);
	if (nolockinode != 0) {
		SMB_STRUCT_STAT tmpsbuf = { .st_ex_ino = nolockinode, };

		ret = fileid_add_nolock_inode(data, &tmpsbuf);
		if (ret != 0) {
			saved_errno = errno;
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}
	}

	if (rootdir_nolock) {
		SMB_STRUCT_STAT rootdirsbuf;

		ret = get_connectpath_ino(handle, ".", &rootdirsbuf);
		if (ret != 0) {
			saved_errno = errno;
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}

		ret = fileid_add_nolock_inode(data, &rootdirsbuf);
		if (ret != 0) {
			saved_errno = errno;
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}
	}

	nolock_paths = lp_parm_string_list(SNUM(handle->conn), "fileid", "nolock_paths", NULL);
	for (i = 0; nolock_paths != NULL && nolock_paths[i] != NULL; i++) {
		SMB_STRUCT_STAT tmpsbuf;

		ret = get_connectpath_ino(handle, nolock_paths[i], &tmpsbuf);
		if (ret == -1 && errno == ENOENT) {
			DBG_ERR("ignoring non existing nolock_paths[%zu]='%s'\n",
				i, nolock_paths[i]);
			continue;
		}
		if (ret != 0) {
			saved_errno = errno;
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}

		ret = fileid_add_nolock_inode(data, &tmpsbuf);
		if (ret != 0) {
			saved_errno = errno;
			SMB_VFS_NEXT_DISCONNECT(handle);
			errno = saved_errno;
			return -1;
		}
		DBG_DEBUG("Adding nolock_paths[%zu]='%s'\n",
			  i, nolock_paths[i]);
	}

	SMB_VFS_HANDLE_SET_DATA(handle, data, NULL,
				struct fileid_handle_data,
				return -1);

	DBG_DEBUG("connect to service[%s] with algorithm[%s] nolock.inodes %zu\n",
		  service, algorithm, data->nolock.num_inodes);

	return 0;
}

static void fileid_disconnect(struct vfs_handle_struct *handle)
{
	const struct loadparm_substitution *lp_sub =
		loadparm_s3_global_substitution();

	DEBUG(10,("fileid_disconnect() connect to service[%s].\n",
		  lp_servicename(talloc_tos(), lp_sub, SNUM(handle->conn))));

	SMB_VFS_NEXT_DISCONNECT(handle);
}

static struct file_id fileid_file_id_create(struct vfs_handle_struct *handle,
					    const SMB_STRUCT_STAT *sbuf)
{
	struct fileid_handle_data *data;
	struct file_id id = { .inode = 0, };

	SMB_VFS_HANDLE_GET_DATA(handle, data,
				struct fileid_handle_data,
				return id);

	id = data->mapping_fn(data, sbuf);
	if (id.extid == 0 && fileid_is_nolock_inode(data, sbuf)) {
		id.extid = data->nolock.extid;
	}

	DBG_DEBUG("Returning dev [%jx] inode [%jx] extid [%jx]\n",
		  (uintmax_t)id.devid, (uintmax_t)id.inode, (uintmax_t)id.extid);

	return id;
}

static struct vfs_fn_pointers vfs_fileid_fns = {
	.connect_fn = fileid_connect,
	.disconnect_fn = fileid_disconnect,
	.file_id_create_fn = fileid_file_id_create
};

static_decl_vfs;
NTSTATUS vfs_fileid_init(TALLOC_CTX *ctx)
{
	NTSTATUS ret;

	ret = smb_register_vfs(SMB_VFS_INTERFACE_VERSION, "fileid",
			       &vfs_fileid_fns);
	if (!NT_STATUS_IS_OK(ret)) {
		return ret;
	}

	vfs_fileid_debug_level = debug_add_class("fileid");
	if (vfs_fileid_debug_level == -1) {
		vfs_fileid_debug_level = DBGC_VFS;
		DEBUG(0, ("vfs_fileid: Couldn't register custom debugging class!\n"));
	} else {
		DEBUG(10, ("vfs_fileid: Debug class number of 'fileid': %d\n", vfs_fileid_debug_level));
	}

	return ret;
}
